SignalR with Open AI: Finish as Sample Guide (#34848)
dotnet/AspNetCore.Docs: aspnetcore/tutorials/ai-powered-group-chat/ai-powered-group-chat.md

This sample guide highlights the process of building a real-time group chat application. In this chat, a group of human collaborators can interact with an AI assistant that has access to the chat history. Any collaborator can invite the AI to assist by starting their message with `@gpt`. The finished app looks like this:

:::image type="content" source="./ai-powered-group-chat.jpg" alt-text="user interface for the AI-powered group chat":::

This sample uses OpenAI for generating intelligent, context-aware responses and SignalR for delivering the response to users in a group. You can find the complete code [in this repo](https://github.com/microsoft/SignalR-Samples-AI/tree/main/AIStreaming).

> ## Dependencies
> Either Azure OpenAI or OpenAI can be used for this project. Make sure to update the `endpoint` and `key` in `appsettings.json`. `OpenAIExtensions` reads the configuration when the app starts, and the configuration values for `endpoint` and `key` are required to authenticate and use either service.

Reviewer: nit: for consistency

```suggestion
## Dependencies

Either Azure OpenAI or OpenAI can be used for this project. Make sure to update the `endpoint` and `key` in `appsettings.json`. `OpenAIExtensions` reads the configuration when the app starts, and the configuration values for `endpoint` and `key` are required to authenticate and use either service.
```

This sample guide highlights the process of building a real-time group chat application. In this chat, a group of human collaborators can interact with an AI assistant that has access to the chat history. Any collaborator can invite the AI to assist by starting their message with `@gpt`. The finished app looks like this:

:::image type="content" source="./ai-powered-group-chat.jpg" alt-text="user interface for the AI-powered group chat":::

This sample uses OpenAI for generating intelligent, context-aware responses and SignalR for delivering the response to users in a group. You can find the complete code [in this repo](https://github.com/microsoft/SignalR-Samples-AI/tree/main/AIStreaming).

## Dependencies

Either Azure OpenAI or OpenAI can be used for this project. Make sure to update the `endpoint` and `key` in `appsettings.json`. `OpenAIExtensions` reads the configuration when the app starts, and the configuration values for `endpoint` and `key` are required to authenticate and use either service.

> # [OpenAI](#tab/open-ai)
> To build this application, the following are required:
> 
> * ASP.NET Core: To create the web application and host the SignalR hub.
> * [SignalR](https://www.nuget.org/packages/Microsoft.AspNetCore.SignalR.Client): For real-time communication between clients, and the server.
> * [OpenAI Client](https://www.nuget.org/packages/OpenAI/2.0.0-beta.10): To interact with OpenAI's API for generating AI responses.
> 
> # [Azure OpenAI](#tab/azure-open-ai)
> To build this application, the following are required:
> 
> * ASP.NET Core: To create the web application and host the SignalR hub.
> * [SignalR](https://www.nuget.org/packages/Microsoft.AspNetCore.SignalR.Client): For real-time communication between clients, and the server.
> * [Azure OpenAI](https://www.nuget.org/packages/Azure.AI.OpenAI/2.0.0-beta.3): Azure.AI.OpenAI
> ---

Reviewer: I believe the added space helps with readability...and not that it really matters, but I've found myself trying to avoid tabs with H1's, you're free to use any heading value so long as the `#tab` bookmark/syntax is in place. Also, removed versions from the packages as stable releases are now available. See https://github.com/microsoft/SignalR-Samples-AI/pull/5, where I just did a PR to upgrade these in the source.

```suggestion
### [OpenAI](#tab/open-ai)

To build this application, you will need the following:
* ASP.NET Core: To create the web application and host the SignalR hub.
* [SignalR](https://www.nuget.org/packages/Microsoft.AspNetCore.SignalR.Client): For real-time communication between clients and the server.
* [OpenAI Client](https://www.nuget.org/packages/OpenAI): To interact with OpenAI's API for generating AI responses.

### [Azure OpenAI](#tab/azure-open-ai)

To build this application, you will need the following:
* ASP.NET Core: To create the web application and host the SignalR hub.
* [SignalR](https://www.nuget.org/packages/Microsoft.AspNetCore.SignalR.Client): For real-time communication between clients and the server.
* [Azure OpenAI](https://www.nuget.org/packages/Azure.AI.OpenAI): `Azure.AI.OpenAI`

---
```

## Implementation

This section highlights the key parts of the code that integrate SignalR with OpenAI to create an AI-enhanced group chat experience.

### Data flow 

The following diagram highlights the step-by-step communication and processing involved in using OpenAI services, employing an iterative approach to responses and data handling:

> :::image type="content" source="./sequence-diagram-ai-powered-group-chat.png" alt-text="sequence diagram for the AI-powered group chat":::

Reviewer: Just an idea, not something that you need to do, but since we still do not have native support from the Learn platform for mermaid diagrams-I've been using mermaid.live and its share feature, that lets you encode/share diagrams. You can build this out, create the share, and embed the share as a comment in the Markdown (which hides it from the end user) but the content developer can use it as a source of truth for future edits, re-exporting it as an image to update the image itself.

For example:

```markdown
<!--

https://mermaid.live/edit#pako:eNp1kk1P4zAQhv_KaK6botLQFnxAyhbt7mFRUQIXlMsQD6lFYmf9wVfV_752UyQQ4NPY7-N3xjPeYmMko0DH_wLrhi8UtZb6WkNcFLzRob9jO-4Hsl41aiDtoQJyUDUblqH7Si-S_lv5P-EOisYroz8zZWKuybYcYx7MZ2KdiLWNaZy35I2FX0F_47ZKbPEaLMPKuN44uPhZH7gKJufnP6AQcG1V27J143kxnpcCVhtuHhyY4EelM2aAS7IP0jwdTPYXJpPnUlQNabiP1dyUfw9OrOUHy7WAK3KO3TtmnaRJqqJkH6x2wI_UBUrvAcsudP5jXatoEmtVzjuQ5GkUJVNswSN5hgIz7Nn2pGQc4jbpNfoN91yjiKGMD6ix1rvIpWlWL7pB4W3gDK0J7QbFPXUu7sIQE7yN_w1hqWLPL8cvsv8pGcZeo9jiM4rl8uh0tjidneXzfDmbH88zfEGRnxxN48rzs3yaLxYn08Uuw1djounx_vbtPk4Jdv8Bo5TJ1A

-->
```

Check out [this link](https://mermaid.live/edit#pako:eNp1kk1P4zAQhv_KaK6botLQFnxAyhbt7mFRUQIXlMsQD6lFYmf9wVfV_752UyQQ4NPY7-N3xjPeYmMko0DH_wLrhi8UtZb6WkNcFLzRob9jO-4Hsl41aiDtoQJyUDUblqH7Si-S_lv5P-EOisYroz8zZWKuybYcYx7MZ2KdiLWNaZy35I2FX0F_47ZKbPEaLMPKuN44uPhZH7gKJufnP6AQcG1V27J143kxnpcCVhtuHhyY4EelM2aAS7IP0jwdTPYXJpPnUlQNabiP1dyUfw9OrOUHy7WAK3KO3TtmnaRJqqJkH6x2wI_UBUrvAcsudP5jXatoEmtVzjuQ5GkUJVNswSN5hgIz7Nn2pGQc4jbpNfoN91yjiKGMD6ix1rvIpWlWL7pB4W3gDK0J7QbFPXUu7sIQE7yN_w1hqWLPL8cvsv8pGcZeo9jiM4rl8uh0tjidneXzfDmbH88zfEGRnxxN48rzs3yaLxYn08Uuw1djounx_vbtPk4Jdv8Bo5TJ1A).

Author: Mermaid: That is really nice, I love it.
